src/client/connection.rs

```diff
@@ -146,13 +144,34 @@ impl Connection {
             Err(e) => {
                 error!("Connection error = {:?}", e);
                 self.handle_connection_error(e);
-                return Err(should_reconnect_again(reconnect_option));
+                return Err(self.should_reconnect_again());
             }
         };
 
         Ok((rt, framed))
     }
 
+    fn should_reconnect_again(&self) -> bool {
+        let reconnect_options = self.mqttoptions.reconnect_opts();
+        let is_disconnecting = self.mqtt_state.clone().borrow().is_disconnecting();
+
+        let reconn_policy_action = match reconnect_options {
+            ReconnectOptions::AfterFirstSuccess(time) => {
+                let time = Duration::from_secs(time);
+                thread::sleep(time);
+                true
+            }
+            ReconnectOptions::Always(time) => {
+                let time = Duration::from_secs(time);
+                thread::sleep(time);
+                true
+            }
+            ReconnectOptions::Never => false,
+        };
+
+        reconn_policy_action && !is_disconnecting
```

Comment:
Nit. Could we do an early return above match statement instead of an `&&` condition? 

```
if state.mqtt_status() == MqttConnectionStatus::Disconnecting {
    return false
}
```